### PR TITLE
[Kernel][UT] Kernel impl. of crossAttnByHead and unit test for cross attention.

### PR DESCRIPTION
**Pull request by pujiang2018**

### CODE REVIEW

File: src/kernels/attention_kernels.h [hunk 1]
```diff
@@ -596,4 +648,135 @@ void crossAttnShardedHead(T1 *output, const T1 *query, const float *attnMask, in
...
+ * @param attnMask Attention mask
+ * @param scale Scale factor for softmax
+ * @param threadNum Thread number
+ * @param getKHead Lambda to get head of cached keys
+ * @param getVHead Lambda to get head of cached values
+ */
+template <typename T, typename KVCacheT, typename Lambda1, typename Lambda2>
+void crossAttnByHead(T *output, const T *query, const T *key, const T *value, int qHeadNum, int kvHeadNum, int headSize,
+        int oStride, int qStride, int kvStride, int batchSize, const int *inputSeqLens, const int *pastSeqLens,
+        bool causal, const float **alibiSlopes, const float **attnMask, const float scale, int threadNum,
```
Comment (abenmao) [inline src/kernels/attention_kernels.h:682]: "const float *" is ok both for alibiSlopes and attnMask

File: src/kernels/attention_kernels.h [hunk 2]
```diff
@@ -596,4 +648,135 @@ void crossAttnShardedHead(T1 *output, const T1 *query, const float *attnMask, in
...
+                int ldc = keyLen;
+                gemmQK(Q, keyMatInfo, S, m, n, headSize, lda, ldc);
+
+                int ldb = kvStride;
+                auto B = key + inputOffsets[b] * kvStride + kvHdx * headSize;
+                small_gemm_transb(Q, B, S + n, m, inputSeqLens[b], headSize, lda, ldb, ldc);
+            }
+
+            // Softmax(Q * K)
+            if (causal && alibiSlopes == nullptr) {
```
Comment (abenmao) [inline src/kernels/attention_kernels.h:741]: During the decoding phase, the softmax computing is the same whether causal is False（chatglm） or True, so “causal” can be omitted in the “if condition”

File: src/kernels/attention_kernels.h [hunk 3]
```diff
@@ -596,4 +648,135 @@ void crossAttnShardedHead(T1 *output, const T1 *query, const float *attnMask, in
...
+                for (int seq = 0; seq < queryLen; ++seq) {
+                    int elements = pastSeqLens[b] + seq + 1;
+                    small_softmax_f32(S + seq * keyLen, scale, elements);
+                    if (keyLen > elements) {
+                        memset(S + seq * keyLen + elements, 0, (keyLen - elements) * sizeof(float));
+                    }
+                }
+            }
+            
+            // TODO: other cases
```
Comment (abenmao) [inline src/kernels/attention_kernels.h:751]: For the Baichuan model, alibiSlopes[i] need to be used in the softmax calculation; I will add a calculation function later.

File: src/kernels/attention_kernels.h (hunk 1 above)
Comment (pujiang2018) [reply src/kernels/attention_kernels.h:682]: Note this is for several samples/sequences, I am not sure if a single pointer is OK.

Comment (abenmao) [reply src/kernels/attention_kernels.h:682]: The alibiSlopes are a set of values that are the same for all sequences. And the AttnMask can also be seen as a large buffer in the bc mode.

Comment (pujiang2018) [reply src/kernels/attention_kernels.h]: Done and removed attnMask.

File: src/kernels/attention_kernels.h (hunk 2 above)
Comment (pujiang2018) [reply src/kernels/attention_kernels.h]: Done.

**Review by abenmao — APPROVED** [review]

**Pull request merged**